fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1155,6 +1155,72 @@ void determineCommittedTransactions(CommitBatchContext* self) {
 	}
 }
 
+inline int64_t extractTenantIdFromKeyRef(StringRef s) {
+	if (s.size() < TenantAPI::PREFIX_SIZE) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	// Parse mutation key to determine tenant prefix
+	StringRef prefix = s.substr(0, TenantAPI::PREFIX_SIZE);
+	return TenantAPI::prefixToId(prefix, EnforceValidTenantId::False);
+}
+
+int64_t extractTenantIdFromSingleKeyMutation(MutationRef m) {
+	ASSERT(!isSystemKey(m.param1));
+
+	// The first 8 bytes of the key of this OP is also an 8-byte number
+	if (m.type == MutationRef::SetVersionstampedKey) {
```

Comment:
Rather than exclude these entirely, we should parse the offset (stored as the last 4 bytes of the key) and determine where the versionstamp will be placed. Only if that versionstamp intersects the first 8 bytes would we disallow it.